File: frontend-service/lib/api/billing.ts
import apiClient from "../api-client"
import type { SelectedFormatPayload } from "./download"

export interface BillingAccountOverviewResponse {
    user_id: string
    currency_code: string
    available_balance_fen: string
    reserved_balance_fen: string
    total_recharged_fen: string
    total_spent_fen: string
    total_traffic_bytes: number
    status: number
    version: number
    created_at: string
    updated_at: string
}

export interface BillingStatementItem {
    statement_id: string
    type: number
    history_id: number
    traffic_bytes: number
    amount_fen: string
    status: number
    remark: string
    created_at: string
}

export interface BillingStatementPageResponse {
    total: number
    page: number
    page_size: number
    items: BillingStatementItem[]
}

export type BillingAccount = BillingAccountOverviewResponse
export type BillingStatementListResponse = BillingStatementPageResponse

export interface BillingEstimateRequest {
    url: string
    platform?: string
    mode?: string
    selected_format?: SelectedFormatPayload
}

export interface BillingEstimateResponse {
    estimated_traffic_bytes: number
    estimated_cost_fen: string
    pricing_version: number
    is_estimated: boolean
    estimate_reason?: string
}

export const billingApi = {
    getAccount: async (): Promise<BillingAccountOverviewResponse> => {
        const response = await apiClient.get("/api/v1/user/account")
        return response.data as BillingAccountOverviewResponse
    },

    listStatements: async (params?: { page?: number; page_size?: number; type?: number; status?: number }): Promise<BillingStatementPageResponse> => {
        const response = await apiClient.get("/api/v1/user/billing/ledger", { params })
        return response.data as BillingStatementPageResponse
    },

    estimateDownload: async (payload: BillingEstimateRequest): Promise<BillingEstimateResponse> => {
        const response = await apiClient.post("/api/v1/user/billing/estimate", payload)
        return response.data as BillingEstimateResponse
    },
}

File: api-gateway/internal/handler/billing.go
package handler

import (
	"context"
	"time"

	"github.com/gin-gonic/gin"

	"youdlp/api-gateway/internal/middleware"
	"youdlp/api-gateway/internal/models"
	pb "youdlp/api-gateway/proto"
)

type BillingHandler struct {
	assetClient pb.AssetServiceClient
	timeout     time.Duration
}

func NewBillingHandler(assetClient pb.AssetServiceClient, timeout time.Duration) *BillingHandler {
	return &BillingHandler{
		assetClient: assetClient,
		timeout:     timeout,
	}
}

func (h *BillingHandler) GetAccount(c *gin.Context) {
	userID := middleware.GetUserID(c)
	if userID == "" {
		models.Unauthorized(c, "user not authenticated")
		return
	}

	ctx, cancel := context.WithTimeout(c.Request.Context(), h.timeout)
	defer cancel()

	resp, err := h.assetClient.GetBillingAccount(ctx, &pb.GetBillingAccountRequest{
		UserId:     userID,
		AutoCreate: true,
	})
	if err != nil {
		writeGRPCError(c, err)
		return
	}

	account := resp.GetAccount()
	models.Success(c, models.BillingAccountOverviewResponse{
		UserID:               account.GetUserId(),
		CurrencyCode:         account.GetCurrencyCode(),
		AvailableBalanceYuan: account.GetAvailableBalanceYuan(),
		ReservedBalanceYuan:  account.GetReservedBalanceYuan(),
		TotalRechargedYuan:   account.GetTotalRechargedYuan(),
		TotalSpentYuan:       account.GetTotalSpentYuan(),
		TotalTrafficBytes:    account.GetTotalTrafficBytes(),
		Status:               account.GetStatus(),
		Version:              account.GetVersion(),
		CreatedAt:            account.GetCreatedAt(),
		UpdatedAt:            account.GetUpdatedAt(),
	})
}

func (h *BillingHandler) ListStatements(c *gin.Context) {
	userID := middleware.GetUserID(c)
	if userID == "" {
		models.Unauthorized(c, "user not authenticated")
		return
	}

	var req models.BillingStatementRequest
	if err := c.ShouldBindQuery(&req); err != nil {
		models.BadRequest(c, "invalid request: "+err.Error())
		return
	}
	if req.Page <= 0 {
		req.Page = 1
	}
	if req.PageSize <= 0 {
		req.PageSize = 20
	}

	ctx, cancel := context.WithTimeout(c.Request.Context(), h.timeout)
	defer cancel()

	resp, err := h.assetClient.ListBillingStatements(ctx, &pb.ListBillingStatementsRequest{
		UserId:   userID,
		Page:     int32(req.Page),
		PageSize: int32(req.PageSize),
		Type:     req.Type,
		Status:   req.Status,
	})
	if err != nil {
		writeGRPCError(c, err)
		return
	}

	items := make([]models.BillingStatementItem, 0, len(resp.GetItems()))
	for _, item := range resp.GetItems() {
		items = append(items, models.BillingStatementItem{
			StatementID:  item.GetStatementId(),
			Type:         item.GetType(),
			HistoryID:    item.GetHistoryId(),
			TrafficBytes: item.GetTrafficBytes(),
			AmountYuan:   item.GetAmountYuan(),
			Status:       item.GetStatus(),
			Remark:       item.GetRemark(),
			CreatedAt:    item.GetCreatedAt(),
		})
	}

	models.Success(c, models.BillingStatementListResponse{
		Total:    resp.GetTotal(),
		Page:     int(resp.GetPage()),
		PageSize: int(resp.GetPageSize()),
		Items:    items,
	})
}

func (h *BillingHandler) Estimate(c *gin.Context) {
	userID := middleware.GetUserID(c)
	if userID == "" {
		models.Unauthorized(c, "user not authenticated")
		return
	}

	var req models.BillingEstimateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		models.BadRequest(c, "invalid request: "+err.Error())
		return
	}

	ctx, cancel := context.WithTimeout(c.Request.Context(), h.timeout)
	defer cancel()

	var selectedFormat *pb.BillingSelectedFormat
	if req.SelectedFormat != nil {
		selectedFormat = &pb.BillingSelectedFormat{
			FormatId:   req.SelectedFormat.FormatID,
			Quality:    req.SelectedFormat.Quality,
			Extension:  req.SelectedFormat.Extension,
			Filesize:   req.SelectedFormat.Filesize,
			Height:     req.SelectedFormat.Height,
			Width:      req.SelectedFormat.Width,
			Fps:        req.SelectedFormat.FPS,
			VideoCodec: req.SelectedFormat.VideoCodec,
			AudioCodec: req.SelectedFormat.AudioCodec,
			Vbr:        req.SelectedFormat.VBR,
			Abr:        req.SelectedFormat.ABR,
			Asr:        req.SelectedFormat.ASR,
		}
	}

	resp, err := h.assetClient.EstimateDownloadBilling(ctx, &pb.EstimateDownloadBillingRequest{
		UserId:         userID,
		Url:            req.URL,
		Platform:       req.Platform,
		Mode:           req.Mode,
		SelectedFormat: selectedFormat,
	})
	if err != nil {
		writeGRPCError(c, err)
		return
	}

	models.Success(c, models.BillingEstimateResponse{
		EstimatedTrafficBytes: resp.GetEstimatedTrafficBytes(),
		EstimatedCostYuan:     resp.GetEstimatedCostYuan(),
		PricingVersion:        resp.GetPricingVersion(),
		IsEstimated:           resp.GetIsEstimated(),
		EstimateReason:        resp.GetEstimateReason(),
	})
}
